front/components/labs/modals/ConfigureLabsConnectionModal.tsx

```diff
@@ -0,0 +1,288 @@
+  SheetContent,
+  SheetFooter,
+  SheetHeader,
+  SheetTitle,
+  SheetTrigger,
+  TrashIcon,
+} from "@dust-tt/sparkle";
+import { useSendNotification } from "@dust-tt/sparkle";
+import type { SetStateAction } from "react";
+import { useState } from "react";
+
+import { DataSourceViewsSpaceSelector } from "@app/components/data_source_view/DataSourceViewsSpaceSelector";
+import type { LabsConnectionsConfigurationResource } from "@app/lib/resources/labs_connections_resource";
+import {
+  useCreateLabsConnectionConfiguration,
+  useDeleteLabsConnectionConfiguration,
+  useLabsConnectionConfiguration,
+  useUpdateLabsConnectionConfiguration,
+} from "@app/lib/swr/labs";
+import type {
+  DataSourceViewSelectionConfigurations,
+  DataSourceViewType,
+  LabsConnectionItemType,
+  LightWorkspaceType,
+  SpaceType,
+} from "@app/types";
+
+interface ConfigureLabsConnectionModal {
+  owner: LightWorkspaceType;
+  connection: LabsConnectionItemType;
+  dataSourcesViews: DataSourceViewType[];
+  spaces: SpaceType[];
+  isSpacesLoading: boolean;
+  existingConfiguration?: LabsConnectionsConfigurationResource;
+}
+
+export function ConfigureLabsConnectionModal({
+  owner,
+  connection,
+  dataSourcesViews,
+  spaces,
+  isSpacesLoading,
+  existingConfiguration,
+}: ConfigureLabsConnectionModal) {
+  const sendNotification = useSendNotification();
+  const updateConnectionConfiguration = useUpdateLabsConnectionConfiguration({
+    workspaceId: owner.sId,
+    connectionId: connection.id,
+  });
+  const createConnectionConfiguration = useCreateLabsConnectionConfiguration({
+    workspaceId: owner.sId,
+  });
+  const deleteConnectionConfiguration = useDeleteLabsConnectionConfiguration({
+    workspaceId: owner.sId,
+    connectionId: connection.id,
+  });
+  const { configuration, mutateConfiguration } = useLabsConnectionConfiguration(
+    {
+      workspaceId: owner.sId,
+      connectionId: connection.id,
+    }
+  );
+
+  const [selectionConfigurations, setSelectionConfigurations] =
+    useState<DataSourceViewSelectionConfigurations>({});
+  const [apiKey, setApiKey] = useState("");
+  const [pendingDataSourceView, setPendingDataSourceView] =
+    useState<DataSourceViewType | null>(null);
+
+  const handleSetConnectionStorageDataSourceView = async (
+    dataSourceView: DataSourceViewType | null
+  ) => {
+    setPendingDataSourceView(dataSourceView);
+  };
+
+  const handleSetSelectionConfigurations = async (
+    newValue: SetStateAction<DataSourceViewSelectionConfigurations>
+  ) => {
+    const newSelectionConfigurations =
+      typeof newValue === "function"
+        ? newValue(selectionConfigurations)
+        : newValue;
+
+    const keys = Object.keys(newSelectionConfigurations);
+
+    if (keys.length === 0) {
+      return;
+    }
+
+    const lastKey = keys[keys.length - 1];
+
+    // If there's no change in the selection, return early
+    if (
+      lastKey &&
+      JSON.stringify(selectionConfigurations[lastKey]) ===
```

Comment:
this I believe is inneficient and a code smell. Is there not a cleaner way to achieve this ? Can track changes and avoid firing if there are no changes ?